qwerty_ast/src/typecheck.rs

```diff
@@ -770,6 +860,128 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
     }
 }
 
+/// Determines if an expression is inherently reversible.
+fn is_expr_inherently_reversible(expr: &Expr, env: &mut TypeEnv) -> Result<bool, TypeError> {
+    match expr {
+        // Operations that break reversibility
+        Expr::Measure { .. } => Ok(false),
+        Expr::Discard { .. } => Ok(false),
+        Expr::Conditional { .. } => Ok(false), // Classical conditionals (if/else) are not allowed in reversible functions
+
+        Expr::Pipe { lhs, rhs, .. } => {
+            let lhs_is_rev = is_expr_inherently_reversible(lhs, env)?;
+            let rhs_is_rev = is_expr_inherently_reversible(rhs, env)?;
+            if !lhs_is_rev || !rhs_is_rev {
+                return Ok(false);
+            }
+
+            let rhs_ty = typecheck_expr(rhs, env)?; // Recursively typecheck to get the actual type of RHS
+            match rhs_ty {
+                Type::RevFuncType { .. } => Ok(true),
+                Type::FuncType { .. } => Ok(false), // Classical functions break reversibility
+                _ => Ok(false),
+            }
+        }
+
+        Expr::Adjoint { func, .. } => {
+            let func_ty = typecheck_expr(func, env)?;
+            match func_ty {
+                Type::RevFuncType { .. } => Ok(true),
+                _ => Ok(false),
+            }
+        }
+
+        Expr::BasisTranslation { .. } => Ok(true),
+
+        Expr::Predicated {
+            then_func,
+            else_func,
+            ..
+        } => {
+            // Predicated operations can be reversible if both their 'then' and 'else' branches are.
+            let then_is_rev = is_expr_inherently_reversible(then_func, env)?;
+            let else_is_rev = is_expr_inherently_reversible(else_func, env)?;
+            Ok(then_is_rev && else_is_rev)
+        }
+
+        Expr::Variable { name, .. } => {
+            if let Some(var_type) = env.get_var(name) {
+                match var_type {
+                    Type::RevFuncType { .. } => Ok(true),
+                    Type::FuncType { .. } => Ok(false), // Classical functions break reversibility
+                    _ => Ok(true), // Non-function vars don't affect reversibility
+                }
+            } else {
+                Ok(true)
+            }
+        }
+
+        // Base cases - inherently reversible
+        Expr::UnitLiteral { .. }
+        | Expr::Tensor { .. }
+        | Expr::NonUniformSuperpos { .. }
+        | Expr::QLit(_) => Ok(true),
+    }
+}
+
+/// Helper function to infer the overall reversibility of a function's body.
+/// This function checks all statements and expressions within the body.
+/// It assumes the function's signature (return type) has already been
+/// validated for consistency with the annotation by `typecheck_function`.
+fn infer_function_body_reversibility(
```

Comment:
I'm wondering if there's a way to integrate this with `typecheck_function()`. For example, maybe `typecheck_function()` could return a `Result<Reversibility, TypeError>` instead of `Result<(), TypeError>`.


I only suggest this `Reversibility` enum is to avoid "Boolean blindness" (if you've heard that term). It could look like this:
```rust
enum Reversiblility {
    NotRev,
    Rev,
}
```

Reply:
Sorry, I don't know why I said `typecheck_function()`. I should have said `typecheck_stmt` and `typecheck_expr()`